From 2fa25e450617cf99fb08111b9003ad4b2dae41cc Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:06:07 -0400
Subject: [ethernet] qlcnic: Fix loopback test failure

Message-id: <1408626422-3941-84-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88676
O-Subject: [PATCH RHEL 7.1 083/138] qlcnic: Fix loopback test failure
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Shahed Shaikh <shahed.shaikh@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id 092dfcf347241576a98d3b1bb4b9b98e0faeb801

Driver was returning from link event handler without
setting linkup variable

Signed-off-by: Shahed Shaikh <shahed.shaikh@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Jitendra Kalsaria <jitendra.kalsaria@qlogic.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_io.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_io.c
index 0ae31f4..b3dc4ec 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_io.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_io.c
@@ -683,12 +683,17 @@ void qlcnic_advert_link_change(struct qlcnic_adapter *adapter, int linkup)
   adapter->ahw->linkup = 0;
   netif_carrier_off(netdev);
  } else if (!adapter->ahw->linkup && linkup) {
-  /* Do not advertise Link up if the port is in loopback mode */
-  if (qlcnic_83xx_check(adapter) && adapter->ahw->lb_mode)
+  adapter->ahw->linkup = 1;
+
+  /* Do not advertise Link up to the stack if device
+   * is in loopback mode
+   */
+  if (qlcnic_83xx_check(adapter) && adapter->ahw->lb_mode) {
+   netdev_info(netdev, "NIC Link is up for loopback test\n");
    return;
+  }
 
   netdev_info(netdev, "NIC Link is up\n");
-  adapter->ahw->linkup = 1;
   netif_carrier_on(netdev);
  }
 }
-- 
1.7.1